CACTUS-346 ActionBar.Panel Component

wilysword opened the pull request:
UAT: https://repayonline.atlassian.net/browse/CACTUS-346

So just implementing this probably wouldn't have taken very long, but I wouldn't be me if I didn't try to overengineer it. At the basic level I wanted to take what I was doing with the ActionBar.Panel and implement the menu hamburger using the same code; but since I knew that _that_ code was also extremely similar to what I had done for submenus on the MenuBar, I wanted to see if I could generalize it further.

The result is the `usePopup` and `useFocusControl` hooks; between the two, I think we could replace a fair bit of the ReachUI stuff if we wanted; though it uses a different strategy with ARIA attributes, so we'll have to double-check how screen readers handle it. (As far as actual implementation I left it at just the hamburger menu, but I still spent a day or two prototyping the submenus to make sure the hooks' interface would work for that use case; sometime in the future I'd like to completely refactor the Menubar using the new hooks.)

A note on the test: I wanted to be more thorough (testing the key handlers and such), but for some reason on some of the tests I would get a console warning about `unstable_flushDiscreteUpdates`. I assume it's something specific to the testing library (like maybe it can't batch the usePopup & useFocusControl state updates together?), because when I did similar actions in the browser I didn't get any errors. Still, keep an eye out for console warnings as you test, and maybe later I'll come back and try to pin the error down more.

mikesoltow commented on modules/cactus-web/src/helpers/usePopup.ts:
```diff
@@ -0,0 +1,186 @@
+  wrapperProps: React.HTMLAttributes<HTMLElement>
+}
+
+function usePopup(
```
If you call this hook with a focus control that only takes one argument and doesn't take the state (and both MenuBar and ActionItemPanel do this, I believe), then calling `toggle` with a non-element `focusHint` does nothing, correct?

wilysword replied:
Depends on the return value. If it takes only one argument and returns undefined, then numeric/text focusHint does nothing. If it returns an array of elements, the focusHint is used to pick an element from that array. (`focusControl` is called [here](https://github.com/repaygithub/cactus/pull/352/files#diff-cdf54dea33530f2901ef38339237c954R107), then the array + focusHint is utilized a few lines below that.)

mikesoltow replied:
Ah gotcha.  I missed its use in `getFocusIndex` somehow.  Thanks!

mikesoltow reviewed (commented):
Everything is working pretty well!  I just have a few comments and questions about the code.

* The width of the panel on full screen mode is small for some reason.  I imagine that you imagine that users would control it with a `width` prop?  Should we set a default `min-width` at certain screen sizes?
* Because `toggle` is passed into the `render` children function, and because `usePopop` is exported, I think we should add some documentation on it (i.e. what is a `focusHint`?  what are the `focusOpts`?  what is a `FocusControl`?).  That would help with the code readability too (although you do admittedly document `FocusControl` in a comment).

mikesoltow reviewed (changes requested):
I don't know what possessed me to do this, but after my initial review, I took a quick look at this on Safari.  Most of it seemed to work well, except the following:

1. Click the gear icon to open the panel.
2. Click the gear icon again to close the panel.  The panel closes briefly, but then opens again.

It looks like the reason is that, in `closeOnBlur`, `relatedTarget` is `undefined` (because why not act more like IE11??).  However, `activeElement` isn't the `button`, like it is in IE, but `body` instead, so `toggle` ends up being called in this handler when it shouldn't be.  :shrug: I hate browsers.

wilysword commented on modules/cactus-web/src/helpers/focus.ts:
```diff
@@ -6,19 +10,169 @@ const FOCUS_SELECTOR =
-    if (el.hasAttribute('tabindex') && el.tabIndex < 0) {
-      return false
+  const result = Array.from(searchFrom.querySelectorAll(FOCUS_SELECTOR)) as HTMLElement[]
+  return result.filter((el) => !(el.hasAttribute('tabindex') && el.tabIndex < 0))
```
It's supposed to. The original purpose of this function is to use in a focus trap, which only applies to elements that are focused by pressing tab. There are certainly use cases where that's not desirable, but it's sufficient as a default.

mikesoltow replied:
Can't you make other elements focusable by adding a `tabIndex` to them?  I can't think of a use case off of the top of my head, but perhaps there are cases where you _want_ to trap focus on a `div` or something.

wilysword replied:
I think you're confused. This only excludes elements with a `tabIndex < 0` (i.e. -1), which are the elements that can be focused by click or script, but not pressing tab.

mikesoltow replied:
Oh I'm definitely confused.  For some reason I thought that the `FOCUS_SELECTOR` wasn't returning something like `<div tabIndex={0} />`, and therefore that not all focusable elements were part of `result`.  But I see that the last line of the selector is `[tabindex]`.  Not sure how I missed that.

Carry on.

wilysword replied:
Yeah, I _really_ wish browsers had a dedicated way to query focusable elements. It's too bad a `:focusable` CSS selector would violate the rules of CSS...

wilysword commented:
> The width of the panel on full screen mode is small for some reason. I imagine that you imagine that users would control it with a width prop? Should we set a default min-width at certain screen sizes?

There's already a min-width (200px), though I suppose we could make it wider; I actually explicitly mention this issue & its cause in the docs, right under where you flagged the typo.

> Because toggle is passed into the render children function, and because usePopop is exported, I think we should add some documentation on it (i.e. what is a focusHint? what are the focusOpts? what is a FocusControl?). That would help with the code readability too (although you do admittedly document FocusControl in a comment).

I agree but I have no idea where to put it or how, since we don't have docs for any of the other helpers and my working knowledge of how to add stuff to the docs is literally just copy/paste. Any pointers would be appreciated.

wilysword commented on modules/cactus-web/src/ActionBar/ActionBar.tsx:
```diff
@@ -11,6 +18,24 @@ interface ItemProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+
+type RenderFn = (t: TogglePopup, expanded: boolean) => React.ReactElement | null
+
+interface PanelProps extends StyleProps, Omit<React.HTMLAttributes<HTMLElement>, 'children'> {
```
It appears you are correct. I was originally trying to use `children` directly as a function, rather than assigning it to another variable, and this was one of my attempts to fix the issues I ran into while doing that. When I gave up on that approach, I never undid it.

mikesoltow replied:
> There's already a min-width (200px), though I suppose we could make it wider; I actually explicitly mention this issue & its cause in the docs, right under where you flagged the typo.

Oops.  Clearly I had forgotten that bit by the time I got here 😄 .  I don't see the `min-width` in either the dev tools or the code (and the panel in the story actually has a width of 105px), but I could be missing something.

> I agree but I have no idea where to put it or how, since we don't have docs for any of the other helpers and my working knowledge of how to add stuff to the docs is literally just copy/paste. Any pointers would be appreciated.

We _could_ look into doing some Gatsby magic to do stuff automatically with co-located `.md` files, but since that would be another ticket, I wonder if, for now, you should create a `Helpers.md` file (or something) inside `docs/Components`, with the goal of spitting out a page that shows up right below Cactus Web > API Docs in the sidebar menu.

wilysword commented on modules/cactus-web/src/ActionBar/ActionBar.tsx:
```diff
@@ -22,17 +47,67 @@ export const ActionBarItem = React.forwardRef<HTMLButtonElement, ItemProps>(
+      }
+    }
+
+    // For some reason Typescript complains about this ref even though it
+    // appears identical to the one in `PanelPopup`, which works just fine...
```
I swear Typescript was complaining about that before, but maybe that was when I was still contemplating supporting an `as` prop. In any case, refs + Typescript = 😢.

wilysword commented on modules/cactus-web/src/ActionBar/ActionBar.tsx:
```diff
@@ -22,17 +47,67 @@ export const ActionBarItem = React.forwardRef<HTMLButtonElement, ItemProps>(
+  }
+)
+
+const PanelPopup = React.forwardRef<HTMLElement, PanelPopupProps>((props, ref) => {
```
This one _does_ support the `as` prop, so I'd prefer to leave it; for example, if I were to try and add an `HTMLUListElement` ref to the PanelPopup in MenuBar, it would complain even though that is in fact the correct type. This way at least people can get the minimum of functionality across all types, and if they need more they can deal with Typescript themselves.

mikesoltow replied:
Ah, good point!  _You_, sir, are the one who is technically correct.

mikesoltow commented:
> There's already a min-width (200px), though I suppose we could make it wider; I actually explicitly mention this issue & its cause in the docs, right under where you flagged the typo.

Oops.  Clearly I had forgotten that bit by the time I got here 😄 .  I don't see the `min-width` in either the dev tools or the code (and the panel in the story actually has a width of 105px), but I could be missing something.

> I agree but I have no idea where to put it or how, since we don't have docs for any of the other helpers and my working knowledge of how to add stuff to the docs is literally just copy/paste. Any pointers would be appreciated.

We _could_ look into doing some Gatsby magic to do stuff automatically with co-located `.md` files, but since that would be another ticket, I wonder if, for now, you should create a `Helpers.md` file (or something) inside `docs/Components`, with the goal of spitting out a page that shows up right below Cactus Web > API Docs in the sidebar menu.

wilysword commented:
I made some changes to how the state updates (`toggle` and `setFocus`) work, so might want to take a look at https://github.com/repaygithub/cactus/pull/352/commits/c8a6c419f0c10da5d43b9deeea27dcd0a6e142c2?w=1.

> Oops. Clearly I had forgotten that bit by the time I got here 😄 . I don't see the min-width in either the dev tools or the code (and the panel in the story actually has a width of 105px), but I could be missing something.

Maybe I removed it and forgot about it 🤦. That sounds like something I might do.

I'll look into the docs thing as soon as I figure out the focus bug. Hopefully I can find a solution to the `relatedTarget` thing once and for all...

Dhalton commented:
Finally had a chance to have a peek at this after, admittedly, putting it off for a couple of days. I didn't come across anything more than what's already been mentioned, though, so I'll come back and give this another look after those issues have been resolved. Great stuff!

wilysword commented:
Added docs for `usePopup`. Let me know if you think anything should be changed, e.g. it's too long, or something needs clarification, or I should document `useFocusControl` separately (and maybe we should export that hook as well?).

mikesoltow reviewed (approved):
No need to change anything with the docs from what I can see!  I don't think it's too long at all, personally: the hook(s) does/do a lot, and this does a great job explaining that IMO.

And I'm fine with not exposing the `useFocusControl` hook for now, as that's pretty low-level.

Thanks a lot for adding that 🎉  🥳

Dhalton commented:
Isn't there supposed to be a min width on the side bar that pops out? In Safari, the side bar has a normal width and is wide enough for the label of the input field to fit on one line, but in Chrome and IE, it's 105px wide and the content gets squished.

wilysword commented:
As brought up in the discussion with Mike, I did originally have a min-width, but I removed it (and then forgot I removed it) because I wasn't sure how to determine what a "good" value would be. Too small and there wouldn't really be any benefit, while too large and the contents might be swimming in a sea of whitespace. (Although I admit that's a somewhat weak argument, since the default could be overridden with the `minWidth` prop.) I figured it'd be better to just recommend that the devs set the width themselves, since they know what the contents are going to be. Still, I'm not opposed to setting a default width or min width, if you think there's a value that works well.

Dhalton commented:
Ah, alright. My interpretation of your conversation with Mike was that you had removed it, forgot you removed it, and were going to add it back. I guess as long as there's mention of the width being configurable in the docs (which it looks like there is) then I'm ok with that. I'll just go ahead and approve this then.

wilysword commented:
Yeah, I guess I never actually explained _why_ I had removed it in that conversation.
